### ISSUE TITLE
onCancel should fire in AlertDialog if Escape is pressed

### ISSUE THREAD

**Issue by kcoop** [adobe/react-spectrum#1773]
# 🐛 Bug Report (or maybe feature request?)

When adding an `onCancel` handler to an AlertDialog, it should fire if the dialog is closed via the Escape key.

See here for a repro: https://codesandbox.io/s/adoring-meadow-7cl4c?file=/src/App.js

Open the dialog, click Cancel, dialog will close, you'll see 'Canceled' in the console.
Open the dialog, press Escape key, dialog will close, no 'Canceled' in the console.

**Comment by snowystinger** [comment]
Yep, I see that. Seems like a bug, but I'll check with the team to make sure it wasn't on purpose. Surprised it didn't come up earlier.

**Comment by snowystinger** [comment]
In the meantime though, you should be able to watch onOpenChange on the DialogTrigger.
I looked into why we don't have it, and the short summary is:
`Dialogs` only fire `onDismiss` for the button in the top right. `Dialogs` also have no idea about the `Esc` key, that’s something that `DialogTrigger` adds to it via `Modal` in `renderOverlay`. So we may need to add this to `DialogTrigger`, which would probably mean a different name for the event. Still talking it over with the team, but figured I'd give an update.

**Comment by tenphi** [comment]
Also, there is no `onDismiss` event while using Dialog component if we click outside the modal or press `Esc`.

**Comment by tenphi** [comment]
As I see even using `onOpenState` there is no difference between the Button that calls `close()` and pressing `Esc`. So I can't tell if a user dismissed the dialog or confirmed it without wrapping `onPress` event for buttons.

**Comment by tenphi** [comment]
Looks like `DialogContainer` fires the `onDismiss` event in both cases. So it would be consistent to have `onDismiss` event on `DialogTrigger` that works the same.

**Comment by snowystinger** [comment]
Adding S2 example https://stackblitz.com/edit/fwp9po8j?file=src%2FExample.tsx

**Comment by arham766** [comment]
Looks like #10136 had an approved approach but was closed over the unsigned CLA. I'd like to pick this up: re-implement that approach with regression tests, and cover the S2 AlertDialog counterpart mentioned in the review as well. Happy to sign the CLA before opening the PR.
